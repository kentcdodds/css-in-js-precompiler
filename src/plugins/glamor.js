import * as glamor from 'glamor'
import {renderStatic} from 'glamor/server'

module.exports = {
  shouldTranspile({code}) {
    return code.indexOf('glamorous') !== -1
  },
  start(callback) {
    let transformed
    const {css, ids} = renderStatic(() => {
      transformed = callback()
      return '<div>fake html to make glamor happy</div>'
    })
    glamor.flush() // make sure multiple runs don't mess things up
    return {transformed, css, ids}
  },
}
