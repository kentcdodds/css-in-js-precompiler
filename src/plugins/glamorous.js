import * as babel from 'babel-core'
import * as glamor from 'glamor'
import {renderStatic} from 'glamor/server'
import {looksLike} from '../utils'

const {types: t} = babel

module.exports = {
  shouldTranspile({code}) {
    return code.indexOf('glamorous') !== -1
  },
  start(callback) {
    let transformed
    const {css, ids} = renderStatic(() => {
      transformed = callback()
      return '<div>fake html to make glamor happy</div>'
    })
    glamor.flush() // make sure multiple runs don't mess things up
    return {transformed, css, ids}
  },
  getImportName(importPath) {
    const defaultSpecifierPath = importPath.get('specifiers')[0]
    if (
      importPath.node.source.value !== 'glamorous' ||
      !t.isImportDefaultSpecifier(defaultSpecifierPath)
    ) {
      return null
    }
    const {node: {local: {name}}} = defaultSpecifierPath
    return name
  },
  getRequireName(variableDeclaratorPath) {
    const {node} = variableDeclaratorPath
    if (!isRequireCall(node.init) || !t.isIdentifier(node.id)) {
      return null
    }
    const {id: {name}} = node
    return name
  },
  getArgumentsPaths(identifierPath) {
    if (
      !looksLike(identifierPath, {
        parentPath: {
          type: type =>
            type === 'MemberExpression' || type === 'CallExpression',
          parentPath: {
            type: 'CallExpression',
          },
        },
      })
    ) {
      return []
    }
    const callExpression = identifierPath.parentPath.parentPath
    return callExpression.get('arguments')
  },
  getClassName(argument) {
    return glamor.css(argument).toString()
  },
}

function isRequireCall(callExpression) {
  return looksLike(callExpression, {
    type: 'CallExpression',
    callee: {
      name: 'require',
    },
    arguments: args =>
      args.length === 1 && looksLike(args[0], {value: 'glamorous'}),
  })
}
