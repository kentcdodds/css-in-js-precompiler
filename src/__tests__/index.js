import fs from 'fs'
import path from 'path'
import cssParser from 'css'
import stripIndent from 'strip-indent'
import * as glamor from 'glamor'
import * as recast from 'recast'

const precompile = require('../')

const babelOptions = {
  parserOpts: {parser: recast.parse},
  generatorOpts: {generator: recast.print, lineTerminator: '\n'},
}

afterEach(() => {
  glamor.flush()
})

const tests = [
  {
    title: 'follows imports',
    fixtureName: 'import.js',
  },
  {
    title: 'follows requires',
    fixtureName: 'require.js',
  },
  {
    title: 'supports statics in arrow functions',
    fixtureName: 'arrow-ternary.js',
  },
  {
    title: 'follows references',
    fixtureName: 'references.js',
  },
  {
    title: 'when creating custom glamorous components',
    fixtureName: 'wrapped-component.js',
  },
  {
    title: 'styles using member expressions',
    fixtureName: 'member-expression-reference.js',
  },
  {
    modifier: 'skip',
    title: 'styles using variables across files',
    fixtureName: 'imported-styles.js',
  },
]

tests.forEach(({title, fixtureName, modifier}, index) => {
  if (modifier) {
    test[modifier](title, testFn)
  } else {
    test(title, testFn)
  }
  function testFn() {
    const sourceFile = path.join(__dirname, '__fixtures__', fixtureName)
    const source = fs.readFileSync(sourceFile, 'utf8')
    const result = precompile({source, babelOptions})

    const {code, css} = result
    const formattedCSS = cssParser.stringify(cssParser.parse(css)).trim()
    const spacer = `\n\n    ðŸ‘‡\n\n`
    const output = `${source.trim()}${spacer}${code.trim()}\n\n${formattedCSS}`

    expect(`\n${output.trim()}\n`).toMatchSnapshot(`${index + 1}. ${title}`)
  }
})

test('does not change code that should not be changed', () => {
  const sourceFile = path.join(__dirname, '__fixtures__/untouched.js')
  const source = fs.readFileSync(sourceFile, 'utf8')
  const result = precompile({sourceFile, babelOptions})

  const {code, css} = result
  expect(code.trim()).toEqual(source.trim())
  expect(css).toEqual('')
})

test('forwards along a bunch of stuff from babel', () => {
  const result = precompile({
    source: stripIndent(
      `
        import glamorous from 'glamorous'
        glamorous.div(
          {
            fontSize: 20,
            fontWeight: 'normal',
          },
          {
            margin: 20,
            ':hover': {
              margin: 10,
            },
          },
        )
        someOtherCall({fontSize: 30})
      `,
    ).trim(),
  })
  expect(Object.keys(result)).toEqual([
    'metadata',
    'options',
    'ignored',
    'code',
    'ast',
    'map',
    'css',
  ])
})
