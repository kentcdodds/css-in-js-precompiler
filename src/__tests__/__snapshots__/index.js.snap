// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. follows imports 1`] = `
"
import fred from 'glamorous'
fred.div({fontSize: 10})

    ðŸ‘‡

import fred from 'glamorous'
fred.div(\\"css-1ea0ufb\\")

.css-1ea0ufb,
[data-css-1ea0ufb] {
  font-size: 10px;
}
"
`;

exports[`2. follows requires 1`] = `
"
const ethel = require('glamorous')
ethel.span({fontSize: 25})

    ðŸ‘‡

const ethel = require('glamorous')
ethel.span(\\"css-1rbjb59\\")

.css-1rbjb59,
[data-css-1rbjb59] {
  font-size: 25px;
}
"
`;

exports[`3. supports statics in arrow functions 1`] = `
"
import glamorous from 'glamorous'

glamorous.div(
  ({big}) =>
    (big
      ? {
          fontSize: 20,
        }
      : {
          fontSize: 12,
        })
)

    ðŸ‘‡

import glamorous from 'glamorous'

glamorous.div(
  ({big}) =>
    (big
      ? \\"css-vfrxuj\\"
      : \\"css-1dmiui7\\")
)

.css-vfrxuj,
[data-css-vfrxuj] {
  font-size: 20px;
}

.css-1dmiui7,
[data-css-1dmiui7] {
  font-size: 12px;
}
"
`;

exports[`4. follows references 1`] = `
"
import glamorous from 'glamorous'

const styles = {color: 'palevioletred'}

glamorous.div(styles)

    ðŸ‘‡

import glamorous from 'glamorous'

const styles = \\"css-zij22n\\"

glamorous.div(styles)

.css-zij22n,
[data-css-zij22n] {
  color: palevioletred;
}
"
`;

exports[`5. when creating custom glamorous components 1`] = `
"
import g from 'glamorous'

g('div')({margin: 2})

    ðŸ‘‡

import g from 'glamorous'

g('div')(\\"css-18570p2\\")

.css-18570p2,
[data-css-18570p2] {
  margin: 2px;
}
"
`;

exports[`6. styles using member expressions 1`] = `
"
import glamorous from 'glamorous'

const colors = {
  primary: 'red',
  secondary: 'blue',
}

glamorous.p({color: colors.primary})

    ðŸ‘‡

import glamorous from 'glamorous'

const colors = {
  primary: 'red',
  secondary: 'blue',
}

glamorous.p(\\"css-1ezp9xe\\")

.css-1ezp9xe,
[data-css-1ezp9xe] {
  color: red;
}
"
`;

exports[`7. styles using variables across files 1`] = `
"
import glamorous from 'glamorous'
import {colors as col} from './styles'

glamorous.p({color: col.primary})

    ðŸ‘‡

import glamorous from 'glamorous'
import {colors as col} from './styles'

glamorous.p(\\"css-1ezp9xe\\")

.css-1ezp9xe,
[data-css-1ezp9xe] {
  color: red;
}
"
`;
